### DISCUSSION TITLE
Qt 6 compatibility

### DISCUSSION THREAD

**Discussion by data-man** [contour-terminal/contour#281] [Ideas]
If you don't mind, I can work on it. :)

**Comment by christianparpart** [comment]
Oh that would be awesome. The only sad story is, I should stay compatible with Ubuntu 18.04 LTS for a little longer. Do you see a possibility to retain this support? (We have a PPA where Ubuntu 18.04 users can apt-get instlal contour from).

@data-man many thanks for your interest, btw, I really appreciate that. :)

EDIT:
I think there is not too much that would need to be touched as I tried to keep the toolkit/windowing dependency as little as possible.

- [TerminalWindow](https://github.com/christianparpart/contour/blob/master/src/contour/TerminalWindow.h): this is the actual window that is being placed on the desktop. you might see some stub code as I already had some PoC tabbed views support in, but that's for later. the basic idea is to be able to have multiple tabs and panes being rendered in one TerminalWindow.
- [opengl/TerminalWidget](https://github.com/christianparpart/contour/blob/master/src/contour/opengl/TerminalWidget.h): that's the QOpenGLWidget and purely bridging between the user interface and the backend APIs (the actual terminal renderer, rendering into pixels is independent of Qt, located in src/terminal_renderer)
- [opengl/OpenGLRenderer](https://github.com/christianparpart/contour/blob/master/src/contour/opengl/OpenGLRenderer.h): doing the actual OpenGL calls
- [Controller](https://github.com/christianparpart/contour/blob/master/src/contour/Controller.h): well yeah, this is currently not really used, the attempt was to be able to show notifications in the system tray, and Qt/Linux support for that was actually already in place (should be revived bevore 0.2.0 elease - for linux at least).

NB: The opengl/TerminalWidget class is an implementation of [TerminalDisplay](https://github.com/christianparpart/contour/blob/master/src/contour/TerminalDisplay.h) in order to allow different display implementations in the future (OpenGL, Software, Vulkan, DirectX, Cocoa).

If you have any questions, just go ahead. We also have a Discord (see README.md) and on Thursdays/Fridays I am also live-streaming working on Contour.

**Comment by data-man** [comment]
> I should stay compatible with Ubuntu 18.04 LTS for a little longer

Just build the release with Qt5 for this OS?

> many thanks for your interest, btw, I really appreciate that. :)

I like your project and someday Contour will be my only terminal. And I really want to get that day closer! :)

Thanks for the tips, I'll explore!

---------------

BTW Maybe create a GH organization (Contour, libunicode, etc.)? (ContourOrg is free, I checked :smile:)

**Comment by christianparpart** [comment]
> > I should stay compatible with Ubuntu 18.04 LTS for a little longer
> 
> Just build the release with Qt5 for this OS?

If the source stays compatible to Qt5, then this is a solution, yeah.

> > many thanks for your interest, btw, I really appreciate that. :)
> 
> I like your project and someday Contour will be my only terminal. And I really want to get that day closer! :)

 💜

> 
> Thanks for the tips, I'll explore!
> 
> BTW Maybe create a GH organization (Contour, libunicode, etc.)? (ContourOrg is free, I checked 😄)

Oh yeah i considered that already. It will come.

EDIT: oh so sad: "contour" as org is reserved by a single dead repository? How can one claim this with an active project. Or we find a better org name

**Comment by data-man** [comment]
> If the source stays compatible to Qt5, then this is a solution, yeah.

I'll explore this later. My tasks now:
- [x] migrate to CPM. It's really cool!
- [ ] CI refactoring. I think testing with MinGW would be useful.

> Oh yeah i considered that already. It will come.

 :100: :+1:

> Or we find a better org name

ContourProjects?

**Reply by christianparpart** [reply]
> CPM

Must work with all existing CI and especially it must still be possible to auto-build packages on PPA (Launchpad) for the Ubuntu versions (18.04, 20.04). It should not be too intrusive by users in order to easily `git clone` and build the project. But I am certainly open to see what CPM can offer us. :)

> CI refactoring.

As long as we stick with Github Actions :) - but having a MingW CI workflow would definitely be interesting. Also, we don't test Windows/Clang builds, only Windows/MSVC, not sure that's adding any gain though. I am currently more interested in getting ARM64 builds into the CI (again).

**Reply by christianparpart** [reply]
> [ ] CI refactoring. I think testing with MinGW would be useful.


Do you actually want to pursue this? (I'm currently walking through all threads) ^^

**Reply by data-man** [reply]
> Do you actually want to pursue this?

Sure!

**Reply by data-man** [reply]
@christianparpart 
Are you using scripts/install-deps.sh for development?
I want to get rid of it.

**Reply by christianparpart** [reply]
Why on earth do you want to get rid of it? It's one easy way to tell people how to install dependencies *easy* when the want to compile from source.

Also, I'm using it locally in fresh VM installs (by coincidence I just did that today on my Ubuntu VM on Windows :D)

**Reply by data-man** [reply]
Got it! Just wondering. :)

**Comment by data-man** [comment]
https://github.com/mitsuba-renderer/nanogui seems interesting, isn't it?

**Reply by christianparpart** [reply]
"NanoGUI builds on GLFW for cross-platform context creation and event handling,". I initially used GLFW and I *had* to move away due to input system issues (which they did acknowledge). Might be interesting to know why Kitty is maintaining its own fork of GLFW inside of Kitty. But that's something I'd like to avoid.

I've check it out (not too deeply) and I do like the look at least. I think it is something to keep an eye on *but* for now I also think we're just fine with Qt (and the move to Qt 6, retaining Qt 5 compat for older distros). I think there is not too much shame in using Qt Widgets library either (or even QML, but I lag experience on that end - eyeing at the command palette feature here).

**Comment by christianparpart** [comment]
We're now live and running at: https://github.com/contour-terminal/contour/ (the old URL still happily exists and just redirects to the org)

**Reply by data-man** [reply]
Thank you!
I couldn't figure out why I had so many (hundreds) notifications. :smiley:

**Reply by christianparpart** [reply]
maybe because I switched all 3 repos into the org? (if that's the notifications u're talking about?)

**Reply by data-man** [reply]
No problem. It's GH bugs. :)

**Comment by data-man** [comment]
Implemented in #311. :beers:
